### ISSUE TITLE
Extension shows as outdated until browser restart

### ISSUE THREAD

**Issue by mikhaelkh** [NeverDecaf/chromium-web-store#4]
[Refined GitHub](https://chrome.google.com/webstore/detail/refined-github/hlepfoohegkhhmjieoechaddaejaokhf) still shows as outdated after I installed update. After browser restart `All extensions are up to date!`

**Comment by NeverDecaf** [comment]
This happens on a handful of extensions and it happens because they are not actually updated until the browser is restarted (guessing due to files in use, etc.) I would like to avoid a band-aid fix such as remembering which extensions were downloaded and hiding them until restart. If anyone can describe a way to detect these "update pending" extensions I will work on a fix.

**Comment by NeverDecaf** [comment]
After looking into it I have made a few discoveries:
- An extension can be notified when it has a pending update if it listens for the [onUpdateAvailable](https://developer.chrome.com/extensions/runtime#event-onUpdateAvailable) event.
- An update will only be installed if the extension's background page is unloaded. This unload can be forced by restarting the browser or calling [chrome.runtime.reload()](https://developer.chrome.com/extensions/runtime#method-reload). 
- **An extension which does not listen for the onUpdateAvailable event will automatically call chrome.runtime.reload() when updated.** Meaning extensions which do not immediately update do so intentionally.
- The info returned by [chrome.management.getAll()](https://developer.chrome.com/extensions/management#method-getAll) will continue to show the outdated version number if an update is pending. Furthermore, [ExtensionInfo](https://developer.chrome.com/extensions/management#type-ExtensionInfo) does not provide information on whether an extension has a pending update or not.

In short, it appears Google does not currently provide a way for one extension to check whether other extensions currently have a pending update.
Additionally, [chrome.management.onInstalled](https://developer.chrome.com/extensions/management#event-onInstalled) is only fired after a pending update is **installed**, not when it is first "queued," meaning a workaround using this method is also impossible.

**Comment by lastweakness** [comment]
> After looking into it I have made a few discoveries:
> 
> * An extension can be notified when it has a pending update if it listens for the [onUpdateAvailable](https://developer.chrome.com/extensions/runtime#event-onUpdateAvailable) event.
> * An update will only be installed if the extension's background page is unloaded. This unload can be forced by restarting the browser or calling [chrome.runtime.reload()](https://developer.chrome.com/extensions/runtime#method-reload).
> * **An extension which does not listen for the onUpdateAvailable event will automatically call chrome.runtime.reload() when updated.** Meaning extensions which do not immediately update do so intentionally.
> * The info returned by [chrome.management.getAll()](https://developer.chrome.com/extensions/management#method-getAll) will continue to show the outdated version number if an update is pending. Furthermore, [ExtensionInfo](https://developer.chrome.com/extensions/management#type-ExtensionInfo) does not provide information on whether an extension has a pending update or not.
> 
> In short, it appears Google does not currently provide a way for one extension to check whether other extensions currently have a pending update.
> Additionally, [chrome.management.onInstalled](https://developer.chrome.com/extensions/management#event-onInstalled) is only fired after a pending update is **installed**, not when it is first "queued," meaning a workaround using this method is also impossible.

So... for now, it's impossible to fix this?

**Comment by shishkin** [comment]
What worked for me with uBlock Origin is to disable the extension in chrome://extensions, do the update this extension, then re-enable uBlock.

Now I'm wondering if Chromium Web Store could disable/enable extensions as it handles updates 🤔

**Comment by NeverDecaf** [comment]
@shishkin Interesting, I believe it is possible to disable/enable other extensions so long as the action originates from a user gesture. I'm not sure exactly how the UI for this would look but I'll do some more investigation.

**Comment by PF4Public** [comment]
@NeverDecaf Have you had some time to investigate since then?

**Comment by NeverDecaf** [comment]
I realized it isn't a good idea to forcibly disable other extensions, so I won't be working on this until google adds a way to query the update status of other extensions within the chrome API.

**Comment by zpackrat** [comment]
Necro'd for input, I've found that if you toggle the Chromium Web Store plugin off and back on it clears the alerts for out of date extensions. Not sure how that impacts the running extensions per the remark about certain plugins only updating on browser restart.

**Comment by username-is-not** [comment]
> Necro'd for input

It freaks me out, when people go off on some tangent... about _resurrecting_ threads, and God-knows-what - like: "Start your own topic". NO, MAN, FFS. :) Topic is >right< there, it's here... subject, content, and what's happening. So, yeah..:) 

_/end rant_
